src/Stack/Setup.hs

```diff
@@ -140,6 +140,8 @@ data SetupOpts = SetupOpts
     , soptsUpgradeCabal :: !Bool
     -- ^ Upgrade the global Cabal library in the database to the newest
     -- version. Only works reliably with a stack-managed installation.
+    , soptsInstallCabal :: !(Maybe Version)
+    -- ^ Install a specific version of Cabal.
```

Comment:
Fusing the two fields and parsing the two options as alternatives would still make sense IMO. As a result the usage info should look roughly like this:

```
stack setup … [--upgrade-cabal | --install-cabal VERSION] …
``` 

The combined field should have a type like `Maybe UpgradeTo` with

```
data UpgradeTo = Latest | Specific Version
```

or so.

The default value (which applies when neither option is used) is `Nothing`.